In [ ]:
from google.colab import files
uploaded = files.upload() # Uploads a file from your local machine

Saving fintrac-canafe_data-donnees.xlsx to fintrac-canafe_data-donnees (4).xlsx


In [ ]:
# Importing necessary packages

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px # Import plotly for interactive plots

pd.set_option('display.max_rows', None) # Set pandas option to display all rows
pd.set_option('display.max_columns', None) # Set pandas option to display all columns

In [ ]:
# Reading file as a dataframe 'df'
# Assumes the uploaded file is named 'fintrac-canafe_data-donnees.xlsx' and contains a sheet named 'DATA'
df = pd.read_excel('fintrac-canafe_data-donnees.xlsx',sheet_name='DATA')

In [ ]:
df.head() # Display the first 5 rows of the DataFrame to get a preview of the data

,ActivitySector_SecteurD'activité,PostalCode_CodePostal,ReportType_TypeDeDéclaration,YearMonthReportReceived_AnnéeMoisDéclarationReçue,NumberOfReports_NombreDeDéclarations,FiscalYear_ExerciceFinancier
0,Accountants_Comptables,_National,LCTR_DOIE,201109,1,2011-2012
1,Accountants_Comptables,_National,LCTR_DOIE,201204,1,2012-2013
2,Accountants_Comptables,_National,LCTR_DOIE,201207,2,2012-2013
3,Accountants_Comptables,_National,LCTR_DOIE,201209,1,2012-2013
4,Accountants_Comptables,_National,LCTR_DOIE,201404,1,2014-2015


In [ ]:
df.info() # Display basic information about the DataFrame, including data types and non-null counts

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 570916 entries, 0 to 570915
Data columns (total 6 columns):
 #   Column                                             Non-Null Count   Dtype 
---  ------                                             --------------   ----- 
 0   ActivitySector_SecteurD'activité                   570916 non-null  object
 1   PostalCode_CodePostal                              570916 non-null  object
 2   ReportType_TypeDeDéclaration                       570916 non-null  object
 3   YearMonthReportReceived_AnnéeMoisDéclarationReçue  570916 non-null  int64 
 4   NumberOfReports_NombreDeDéclarations               570916 non-null  int64 
 5   FiscalYear_ExerciceFinancier                       570916 non-null  object
dtypes: int64(2), object(4)
memory usage: 26.1+ MB


In [ ]:
duplicate_rows = df.duplicated().sum() # Calculate the number of duplicate rows
print(f"Number of duplicate rows: {duplicate_rows}") # Print the number of duplicate rows

Number of duplicate rows: 0


In [ ]:
df.columns # Display the column names of the DataFrame

Index(['ActivitySector_SecteurD'activité', 'PostalCode_CodePostal',
       'ReportType_TypeDeDéclaration',
       'YearMonthReportReceived_AnnéeMoisDéclarationReçue',
       'NumberOfReports_NombreDeDéclarations', 'FiscalYear_ExerciceFinancier'],
      dtype='object')

In [ ]:
# Update the column names by splitting at '_' and keeping the first part (English part)
df.columns = df.columns.str.split('_').str[0]
display(df.columns) # Display the updated column names

Index(['ActivitySector', 'PostalCode', 'ReportType', 'YearMonthReportReceived',
       'NumberOfReports', 'FiscalYear'],
      dtype='object')

In [ ]:
# Update each row in object type columns by splitting at '_' and keeping the first part (English part)
for col in df.columns:
    if df[col].dtype == 'object':
        df[col] = df[col].astype(str).str.split('_').str[0]

display(df.head()) # Display the first few rows of the DataFrame with updated data

,ActivitySector,PostalCode,ReportType,YearMonthReportReceived,NumberOfReports,FiscalYear
0,Accountants,,LCTR,201109,1,2011-2012
1,Accountants,,LCTR,201204,1,2012-2013
2,Accountants,,LCTR,201207,2,2012-2013
3,Accountants,,LCTR,201209,1,2012-2013
4,Accountants,,LCTR,201404,1,2014-2015


In [ ]:
df.head() # Display the first few rows of the DataFrame again

,ActivitySector,PostalCode,ReportType,YearMonthReportReceived,NumberOfReports,FiscalYear
0,Accountants,,LCTR,201109,1,2011-2012
1,Accountants,,LCTR,201204,1,2012-2013
2,Accountants,,LCTR,201207,2,2012-2013
3,Accountants,,LCTR,201209,1,2012-2013
4,Accountants,,LCTR,201404,1,2014-2015


In [ ]:
blank_postal_codes = (df['PostalCode'] == '').sum() # Count the number of rows where 'PostalCode' is blank
print(f"Number of blank postal codes: {blank_postal_codes}") # Print the count of blank postal codes

Number of blank postal codes: 3459


In [ ]:
# Convert 'YearMonthReportReceived' column to string type (already done in previous cleaning step, can be removed if confident in order)
df['YearMonthReportReceived'] = df['YearMonthReportReceived'].astype(str)

In [ ]:
# Remove rows where 'PostalCode' is an empty string and create a copy of the filtered DataFrame
df = df[df['PostalCode'] != ''].copy()
display(df.head()) # Display the first few rows of the DataFrame after removing rows with blank postal codes

,ActivitySector,PostalCode,ReportType,YearMonthReportReceived,NumberOfReports,FiscalYear
27,Accountants,T,LCTR,201510,7,2015-2016
28,Accountants,T,LCTR,201511,5,2015-2016
29,Accountants,T5,LCTR,201510,7,2015-2016
30,Accountants,T5,LCTR,201511,5,2015-2016
31,Accountants,T5J,LCTR,201510,7,2015-2016


In [ ]:
# Separate Year and Month from 'YearMonthReportReceived' column
df['YearReportReceived'] = df['YearMonthReportReceived'].astype(str).str[:4] # Extract the first 4 characters as Year
df['MonthReportReceived'] = df['YearMonthReportReceived'].astype(str).str[4:] # Extract characters from the 5th position onwards as Month
display(df.head()) # Display the first few rows with the new 'YearReportReceived' and 'MonthReportReceived' columns

,ActivitySector,PostalCode,ReportType,YearMonthReportReceived,NumberOfReports,FiscalYear,YearReportReceived,MonthReportReceived
27,Accountants,T,LCTR,201510,7,2015-2016,2015,10
28,Accountants,T,LCTR,201511,5,2015-2016,2015,11
29,Accountants,T5,LCTR,201510,7,2015-2016,2015,10
30,Accountants,T5,LCTR,201511,5,2015-2016,2015,11
31,Accountants,T5J,LCTR,201510,7,2015-2016,2015,10


In [ ]:
df.ActivitySector.unique() # Get the unique values in the 'ActivitySector' column

array(['Accountants', 'Banks', 'BC Notaries',
       'Credit Unions / Caisses populaires',
       'Dealer in precious metals and stones', 'Life Insurance',
       'Money services businesses', 'Real Estate', 'Securities Dealers',
       'Trust & Loan companies'], dtype=object)

In [ ]:
# Remove the French part from 'ActivitySector' values by splitting at ' /' and keeping the first part
df['ActivitySector'] = df['ActivitySector'].str.split(' /').str[0]

In [ ]:
df['ActivitySector'].unique() # Display the unique values in the 'ActivitySector' column after cleaning

array(['Accountants', 'Banks', 'BC Notaries', 'Credit Unions',
       'Dealer in precious metals and stones', 'Life Insurance',
       'Money services businesses', 'Real Estate', 'Securities Dealers',
       'Trust & Loan companies'], dtype=object)

In [ ]:
display(df['ReportType'].unique()) # Display the unique values in the 'ReportType' column

array(['LCTR', 'EFT', 'STR'], dtype=object)

These abbreviations stands for:

LCTR (Large Cash Transaction Reports)

EFT (Electronic Funds Transfer Reports)

STR (Suspicious Transaction Reports)

CDR (Casino Disbursement Reports)

In [ ]:
# Convert 'YearMonthReportReceived' to datetime objects for time series analysis
df['YearMonthReportReceived'] = pd.to_datetime(df['YearMonthReportReceived'], format='%Y%m')

# Group by year and month and sum the number of reports to get the trend over time
reports_over_time = df.groupby('YearMonthReportReceived')['NumberOfReports'].sum().reset_index()

# Create an interactive line plot using Plotly Express to visualize the trend
fig = px.line(reports_over_time, x='YearMonthReportReceived', y='NumberOfReports', title='Trend of Number of Reports Over Time')
fig.update_layout(xaxis_title='Date', yaxis_title='Number of Reports')
fig.update_xaxes(dtick="M12", tickformat="%Y") # Set x-axis ticks to show each year
fig.show()

In [ ]:
# Calculate monthly percentage change in the number of reports
reports_over_time['PercentageChange'] = reports_over_time['NumberOfReports'].pct_change() * 100
reports_over_time['Year'] = reports_over_time['YearMonthReportReceived'].dt.year # Extract Year

# Group by year and sum the number of reports to get yearly totals
yearly_reports = reports_over_time.groupby('Year')['NumberOfReports'].sum().reset_index()
# Calculate yearly percentage change in the number of reports
yearly_reports['YearlyPercentageChange'] = yearly_reports['NumberOfReports'].pct_change() * 100

print("Monthly Percentage Change in Number of Reports:")
display(reports_over_time[['YearMonthReportReceived', 'NumberOfReports', 'PercentageChange']]) # Display monthly changes

print("\nYearly Percentage Change in Number of Reports:")
display(yearly_reports) # Display yearly changes

Monthly Percentage Change in Number of Reports:


,YearMonthReportReceived,NumberOfReports,PercentageChange
0,2011-04-01,4105337,NaN
1,2011-05-01,4303022,4.815317
2,2011-06-01,4477522,4.055290
3,2011-07-01,4320281,-3.511786
4,2011-08-01,4641816,7.442456
5,2011-09-01,4298693,-7.391999
6,2011-10-01,4186439,-2.611352
7,2011-11-01,4367417,4.322958
8,2011-12-01,4779224,9.429074
9,2012-01-01,4428115,-7.346569



Yearly Percentage Change in Number of Reports:


,Year,NumberOfReports,YearlyPercentageChange
0,2011,39479751,NaN
1,2012,57586112,45.862399
2,2013,57598118,0.020849
3,2014,59763276,3.759078
4,2015,66523318,11.311365
5,2016,70919919,6.609113
6,2017,73870885,4.160983
7,2018,80380916,8.812716
8,2019,89878853,11.816159
9,2020,88615870,-1.405206


**Trend of Reports Over Time:**

To quantify the observed trend, we calculated the percentage change in the number of reports:

*   **Monthly Percentage Change:** We can see month-over-month fluctuations in the number of reports.
*   **Yearly Percentage Change:** The yearly percentage change shows the overall growth or decline in the number of reports each year. From the data, there's a significant yearly increase in reports, particularly between 2011 and 2012 (45.86%). While there are fluctuations, the trend generally indicates an increase in the number of reports over the years, with a substantial increase of 20.44% between 2022 and 2023.

In [ ]:
# Export the cleaned DataFrame to a CSV file
df.to_csv('fintrac_data_cleaned.csv', index=False)
print("Updated data exported to fintrac_data_cleaned.csv")
#files.download('fintrac_data_cleaned.csv') # Code to download the file (commented out)

Updated data exported to fintrac_data_cleaned.csv


In [ ]:
# Export the cleaned DataFrame to an Excel file
df.to_excel('fintrac_data_cleaned.xlsx', index=False)
print("Updated data exported to fintrac_data_cleaned.xlsx")

#files.download('fintrac_data_cleaned.xlsx') # Code to download the file (commented out)

Updated data exported to fintrac_data_cleaned.xlsx



Group the dataframe by year and activity sector and sum the number of reports to analyze the yearly trend by sector.



In [ ]:
# Group by year and activity sector and sum the number of reports
yearly_sector_reports = df.groupby(['YearReportReceived', 'ActivitySector'])['NumberOfReports'].sum().reset_index()
display(yearly_sector_reports) # Display the aggregated data

,YearReportReceived,ActivitySector,NumberOfReports
0,2011,Banks,37098147
1,2011,Credit Unions,1510028
2,2011,Dealer in precious metals and stones,1222
3,2011,Life Insurance,760
4,2011,Money services businesses,755215
5,2011,Securities Dealers,1508
6,2011,Trust & Loan companies,112871
7,2012,Banks,53621530
8,2012,Credit Unions,1923460
9,2012,Dealer in precious metals and stones,1941


## Find the highest contributor per year

For each year, identify the Activity Sector with the maximum number of reports.


In [ ]:
# Find the activity sector with the maximum number of reports for each year
# .loc[yearly_sector_reports.groupby('YearReportReceived')['NumberOfReports'].idxmax()] gets the rows with the max reports per year
highest_contributor_per_year = yearly_sector_reports.loc[yearly_sector_reports.groupby('YearReportReceived')['NumberOfReports'].idxmax()]
display(highest_contributor_per_year) # Display the highest contributing sector per year

,YearReportReceived,ActivitySector,NumberOfReports
0,2011,Banks,37098147
7,2012,Banks,53621530
14,2013,Banks,53459402
21,2014,Banks,55641252
29,2015,Banks,61871451
36,2016,Banks,66349456
45,2017,Banks,68186266
53,2018,Banks,73816152
61,2019,Banks,82451330
69,2020,Banks,81663789


In [ ]:
# Create a line plot to visualize the yearly trend of reports by activity sector
fig = px.line(yearly_sector_reports, x='YearReportReceived', y='NumberOfReports', color='ActivitySector', title='Yearly Trend of Number of Reports by Activity Sector')
fig.update_layout(xaxis_title='Year', yaxis_title='Number of Reports')
fig.show() # Display the plot

In [ ]:
# Filter out the 'Banks' sector to better visualize trends in other sectors
yearly_sector_reports_filtered = yearly_sector_reports[yearly_sector_reports['ActivitySector'] != 'Banks'].copy()

# Create a line plot for yearly trends excluding the 'Banks' sector
fig = px.line(yearly_sector_reports_filtered, x='YearReportReceived', y='NumberOfReports', color='ActivitySector', title='Yearly Trend of Reports by Activity Sector (Excluding Banks)')
fig.update_layout(xaxis_title='Year', yaxis_title='Number of Reports')
fig.show() # Display the plot

In [ ]:
banks_df = df[df['ActivitySector'] == 'Banks'].copy() # Filter the DataFrame to include only rows where 'ActivitySector' is 'Banks'
display(banks_df.head()) # Display the first few rows of the Banks DataFrame

,ActivitySector,PostalCode,ReportType,YearMonthReportReceived,NumberOfReports,FiscalYear,YearReportReceived,MonthReportReceived
492,Banks,A,EFT,2011-04-01,211,2011-2012,2011,04
493,Banks,A,EFT,2011-05-01,240,2011-2012,2011,05
494,Banks,A,EFT,2011-06-01,182,2011-2012,2011,06
495,Banks,A,EFT,2011-07-01,187,2011-2012,2011,07
496,Banks,A,EFT,2011-08-01,197,2011-2012,2011,08


In [ ]:
# Group by year and report type and sum the number of reports
report_type_yearly = df.groupby(['YearReportReceived', 'ReportType'])['NumberOfReports'].sum().reset_index()

# Pivot the table to have years as index, report types as columns, and sum of reports as values
report_type_yearly_pivot = report_type_yearly.pivot_table(index='YearReportReceived', columns='ReportType', values='NumberOfReports', fill_value=0)

display(report_type_yearly_pivot) # Display the pivoted table

ReportType,EFT,LCTR,STR
YearReportReceived,,,
2011,22432432.0,16923653.0,123666.0
2012,32917396.0,24494928.0,173788.0
2013,33291853.0,24094472.0,211793.0
2014,35506532.0,24048465.0,208279.0
2015,41255467.0,24998445.0,269406.0
2016,40173633.0,30440556.0,305730.0
2017,43861421.0,29595367.0,414097.0
2018,50452905.0,29364028.0,563983.0
2019,59952302.0,28890155.0,1036396.0


In [ ]:
# Create a stacked bar chart to visualize the yearly distribution of report types
fig = px.bar(report_type_yearly, x='YearReportReceived', y='NumberOfReports', color='ReportType',
             title='Yearly Distribution of Report Types', labels={'YearReportReceived': 'Year', 'NumberOfReports': 'Number of Reports'})
fig.update_layout(xaxis_title='Year', yaxis_title='Number of Reports')
fig.show() # Display the chart

In [ ]:
# Group by PostalCode and sum the number of reports to analyze geographical distribution
reports_by_postal_code = df.groupby('PostalCode')['NumberOfReports'].sum().reset_index()
reports_by_postal_code = reports_by_postal_code.sort_values(by='NumberOfReports', ascending=False) # Sort by number of reports in descending order

print("Top 10 Postal Codes by Number of Reports:")
display(reports_by_postal_code.head(10)) # Display the top 10 postal codes

Top 10 Postal Codes by Number of Reports:


,PostalCode,NumberOfReports
650,M,203311912
707,M5,194678055
714,M5J,95326275
201,H,40239203
715,M5K,37460734
712,M5G,25375278
1163,V,22042488
220,H2,19309453
494,L,18780957
239,H2Y,18711649


In [ ]:
# Function to map the first letter of a postal code to a province/territory
def get_province(postal_code):
    if pd.isna(postal_code) or postal_code == '':
        return 'Unknown'
    first_letter = str(postal_code)[0].upper()
    if first_letter == 'T':
        return 'Alberta'
    elif first_letter == 'V':
        return 'British Columbia'
    elif first_letter == 'R':
        return 'Manitoba'
    elif first_letter == 'E':
        return 'New Brunswick'
    elif first_letter == 'A':
        return 'Newfoundland and Labrador'
    elif first_letter == 'B':
        return 'Nova Scotia'
    elif first_letter in ['K', 'L', 'M', 'N', 'P']:
        return 'Ontario'
    elif first_letter == 'C':
        return 'Prince Edward Island'
    elif first_letter in ['G', 'H', 'J']: # K is also listed for Ontario and Quebec, need clarification or handle potential overlap
         return 'Quebec'
    elif first_letter == 'S':
        return 'Saskatchewan'
    elif first_letter == 'X':
        return 'Territories (NT/NU)' # Cannot distinguish between NT and NU with only the first letter
    elif first_letter == 'Y':
        return 'Yukon'
    else:
        return 'Unknown'

df['Province_Territory'] = df['PostalCode'].apply(get_province) # Apply the function to create a new 'Province_Territory' column
display(df.head()) # Display the first few rows with the new column

,ActivitySector,PostalCode,ReportType,YearMonthReportReceived,NumberOfReports,FiscalYear,YearReportReceived,MonthReportReceived,Province_Territory
27,Accountants,T,LCTR,2015-10-01,7,2015-2016,2015,10,Alberta
28,Accountants,T,LCTR,2015-11-01,5,2015-2016,2015,11,Alberta
29,Accountants,T5,LCTR,2015-10-01,7,2015-2016,2015,10,Alberta
30,Accountants,T5,LCTR,2015-11-01,5,2015-2016,2015,11,Alberta
31,Accountants,T5J,LCTR,2015-10-01,7,2015-2016,2015,10,Alberta


In [ ]:
# Group by Province_Territory and sum the number of reports
reports_by_province = df.groupby('Province_Territory')['NumberOfReports'].sum().reset_index()
reports_by_province = reports_by_province.sort_values(by='NumberOfReports', ascending=False) # Sort by number of reports

print("Total Number of Reports by Province/Territory:")
display(reports_by_province) # Display the total reports by province/territory

Total Number of Reports by Province/Territory:


,Province_Territory,NumberOfReports
6,Ontario,709836041
8,Quebec,149246380
1,British Columbia,64765844
0,Alberta,39567803
5,Nova Scotia,12252826
2,Manitoba,12046053
9,Saskatchewan,9401449
3,New Brunswick,6687069
4,Newfoundland and Labrador,4088897
7,Prince Edward Island,1353449


In [ ]:
# Create a bar chart to visualize the total number of reports by province/territory
fig = px.bar(reports_by_province, x='Province_Territory', y='NumberOfReports',
             title='Total Number of Reports by Province/Territory',
             labels={'Province_Territory': 'Province/Territory', 'NumberOfReports': 'Number of Reports'},
             text=reports_by_province['NumberOfReports'].apply(lambda x: f'{x/1000000:.1f}M')) # Format text labels in millions on top of bars
fig.update_layout(xaxis_title='Province/Territory', yaxis_title='Number of Reports', template='plotly') # Set layout and theme
fig.show() # Display the chart

In [ ]:
# Group by year and province/territory and sum the number of reports for yearly distribution
reports_yearly_province = df.groupby(['YearReportReceived', 'Province_Territory'])['NumberOfReports'].sum().reset_index()

# Create a stacked bar chart to visualize the yearly distribution of reports by province/territory
fig = px.bar(reports_yearly_province, x='YearReportReceived', y='NumberOfReports', color='Province_Territory',
             title='Yearly Distribution of Reports by Province/Territory',
             labels={'YearReportReceived': 'Year', 'NumberOfReports': 'Number of Reports', 'Province_Territory': 'Province/Territory'})
fig.update_layout(xaxis_title='Year', yaxis_title='Number of Reports')
fig.show() # Display the chart

In [ ]:
reports_yearly_province = df.groupby(['YearReportReceived', 'Province_Territory'])['NumberOfReports'].sum().reset_index()

# Use apply with nlargest on each group to find top 3 provinces per year
top_provinces_yearly = reports_yearly_province.groupby('YearReportReceived').apply(lambda x: x.nlargest(3, 'NumberOfReports')).reset_index(drop=True)

display(top_provinces_yearly) # Display the top 3 provinces per year

/tmp/ipython-input-2109175914.py:4: DeprecationWarning:

DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.



,YearReportReceived,Province_Territory,NumberOfReports
0,2011,Ontario,26208460
1,2011,Quebec,5943414
2,2011,British Columbia,2733606
3,2012,Ontario,37572363
4,2012,Quebec,8489246
5,2012,British Columbia,4580999
6,2013,Ontario,37141455
7,2013,Quebec,8656573
8,2013,British Columbia,4834164
9,2014,Ontario,38622841


In [ ]:
# Create a grouped bar chart to visualize the top 3 contributing provinces/territories by year
fig = px.bar(top_provinces_yearly, x='YearReportReceived', y='NumberOfReports', color='Province_Territory',
             title='Top 3 Contributing Provinces/Territories by Year',
             labels={'YearReportReceived': 'Year', 'NumberOfReports': 'Number of Reports', 'Province_Territory': 'Province/Territory'},
             barmode='group') # Use barmode='group' for grouped bar chart
fig.update_layout(xaxis_title='Year', yaxis_title='Number of Reports')
fig.show() # Display the chart

In [ ]:
# Group by year, province/territory, and report type and sum reports for detailed distribution
reports_yearly_province_type = df.groupby(['YearReportReceived', 'Province_Territory', 'ReportType'])['NumberOfReports'].sum().reset_index()

fig = px.bar(reports_yearly_province_type, x='YearReportReceived', y='NumberOfReports', color='ReportType',
             title='Yearly Distribution of Report Types by Province/Territory',
             labels={'YearReportReceived': 'Year', 'NumberOfReports': 'Number of Reports', 'ReportType': 'Report Type'},
             facet_col='Province_Territory', facet_col_wrap=4,
             facet_row_spacing=0.05, facet_col_spacing=0.05, # Adjust spacing between facets
             height=800) # Increase height for better readability

fig.update_layout(xaxis_title='Year', yaxis_title='Number of Reports')
fig.for_each_annotation(lambda a: a.update(text=a.text.split("=")[-1])) # Clean up facet titles

# Update y-axis to be independent for each facet to better visualize trends in provinces with fewer reports
fig.update_yaxes(matches=None)
fig.show() # Display the faceted chart

In [ ]:
# Filter for only STR reports
str_reports_yearly_province = reports_yearly_province_type[reports_yearly_province_type['ReportType'] == 'STR']

# Create a faceted bar chart for yearly distribution of STR reports by province/territory
fig = px.bar(str_reports_yearly_province, x='YearReportReceived', y='NumberOfReports',
             title='Yearly Distribution of STR Reports by Province/Territory',
             labels={'YearReportReceived': 'Year', 'NumberOfReports': 'Number of Reports', 'ReportType': 'Report Type'},
             facet_col='Province_Territory', facet_col_wrap=4,
             facet_row_spacing=0.05, facet_col_spacing=0.05, # Adjust spacing between facets
             height=800,
             color_discrete_sequence=['green']) # Set a fixed color to green for STR reports

fig.update_layout(xaxis_title='Year', yaxis_title='Number of Reports')
fig.for_each_annotation(lambda a: a.update(text=a.text.split("=")[-1])) # Clean up facet titles

# Update y-axis to be independent for each facet
fig.update_yaxes(matches=None)
fig.show() # Display the faceted STR chart

In [ ]:
# Chart for LCTR reports in red
lctr_reports_yearly_province = reports_yearly_province_type[reports_yearly_province_type['ReportType'] == 'LCTR']

fig_lctr = px.bar(lctr_reports_yearly_province, x='YearReportReceived', y='NumberOfReports',
                  title='Yearly Distribution of LCTR Reports by Province/Territory',
                  labels={'YearReportReceived': 'Year', 'NumberOfReports': 'Number of Reports', 'ReportType': 'Report Type'},
                  facet_col='Province_Territory', facet_col_wrap=4,
                  facet_row_spacing=0.05, facet_col_spacing=0.05,
                  height=800,
                  color_discrete_sequence=['red']) # Set color to red for LCTR reports

fig_lctr.update_layout(xaxis_title='Year', yaxis_title='Number of Reports')
fig_lctr.for_each_annotation(lambda a: a.update(text=a.text.split("=")[-1]))
fig_lctr.update_yaxes(matches=None)
fig_lctr.show() # Display the faceted LCTR chart

In [ ]:
# Chart for EFT reports in blue
eft_reports_yearly_province = reports_yearly_province_type[reports_yearly_province_type['ReportType'] == 'EFT']

fig_eft = px.bar(eft_reports_yearly_province, x='YearReportReceived', y='NumberOfReports',
                 title='Yearly Distribution of EFT Reports by Province/Territory',
                 labels={'YearReportReceived': 'Year', 'NumberOfReports': 'Number of Reports', 'ReportType': 'Report Type'},
                 facet_col='Province_Territory', facet_col_wrap=4,
                 facet_row_spacing=0.05, facet_col_spacing=0.05,
                 height=800,
                 color_discrete_sequence=['blue']) # Set color to blue for EFT reports

fig_eft.update_layout(xaxis_title='Year', yaxis_title='Number of Reports')
fig_eft.for_each_annotation(lambda a: a.update(text=a.text.split("=")[-1]))
fig_eft.update_yaxes(matches=None)
fig_eft.show() # Display the faceted EFT chart

In [ ]:
# Group by FiscalYear and sum the number of reports
reports_by_fiscal_year = df.groupby('FiscalYear')['NumberOfReports'].sum().reset_index()

# Create a bar chart to visualize the total number of reports by fiscal year
fig = px.bar(reports_by_fiscal_year, x='FiscalYear', y='NumberOfReports',
             title='Total Number of Reports by Fiscal Year',
             labels={'FiscalYear': 'Fiscal Year', 'NumberOfReports': 'Number of Reports'},
             text=reports_by_fiscal_year['NumberOfReports'].apply(lambda x: f'{x/1000000:.1f}M')) # Format text labels in millions on top of bars
fig.update_layout(xaxis_title='Fiscal Year', yaxis_title='Number of Reports')
fig.show() # Display the chart

# FINTRAC Data Analysis

This notebook provides an analysis of FINTRAC (Financial Transactions and Reports Analysis Centre of Canada) data. The analysis focuses on understanding trends in the number of reports submitted over time, the geographical distribution of these reports across Canada, and the breakdown of report types within different activity sectors.

## Notebook Structure and Analysis Steps

1.  **Data Loading and Initial Inspection**: The analysis begins by loading the data from an Excel file (`fintrac-canafe_data-donnees.xlsx`) into a pandas DataFrame. Initial steps include previewing the data, checking data types, and identifying duplicate rows.
2.  **Data Cleaning and Preparation**: This section focuses on cleaning the data for analysis. This involves:
    *   Cleaning column names to retain only the English portion.
    *   Cleaning data within columns (object types) to remove French text.
    *   Handling missing or blank postal codes by removing relevant rows.
    *   Extracting Year and Month into separate columns from the `YearMonthReportReceived` column.
    *   Cleaning the 'ActivitySector' column.
3.  **Trend Analysis**: This part of the analysis examines how the number of reports has changed over time.
    *   The total number of reports is aggregated by year and month.
    *   An interactive line plot visualizes the trend of reports over time.
    *   Monthly and yearly percentage changes in report numbers are calculated and displayed to quantify the trend.
4.  **Geographical Distribution Analysis**: This section explores where the reports are originating from within Canada.
    *   The total number of reports is aggregated by Postal Code to identify areas with high reporting volumes.
    *   A new column `Province_Territory` is created based on the first letter of the Postal Code.
    *   The total number of reports is aggregated and visualized by `Province_Territory` using a bar chart.
    *   The top 3 contributing provinces/territories by year are identified and visualized.
5.  **Yearly Distribution by Province and Report Type**: A detailed breakdown of report types within each province/territory over the years is performed and visualized using faceted bar charts for all report types, and then specifically for STR, LCTR, and EFT reports.
6.  **Fiscal Year Trend**: The total number of reports is analyzed and visualized by Fiscal Year.
7.  **Cross-analysis by Activity Sector and Report Type**: The analysis attempts to explore the distribution of report types within different activity sectors, noting the dominance of the 'Banks' sector.

## Data Sources:

*   **FINTRAC Data on Financial Transactions:**
    *   **Dataset:** Financial transaction report counts by postal code and activity sector
    *   **Details:** This dataset provides monthly counts of financial transaction reports submitted to FINTRAC, categorized by activity sector, report type, and reporting entity location.
    *   **Usage:** Use this data to analyze the volume and distribution of various financial transactions across different sectors and regions.
*   **FINTRAC Publications and Reports:**
    *   **Source:** FINTRAC Publications
    *   **Details:** FINTRAC publishes strategic intelligence reports, operational alerts, and sectoral advisories that can provide insights into emerging trends and typologies in money laundering activities.
    *   **Usage:** Incorporate findings from these publications to contextualize your dashboard and highlight areas of concern.

## Key Observations and Findings

📊 **Here’s what the data reveals:**

🔼 1.35+ billion reports were filed between 2011 and 2023, marking a ~70% increase in total reporting volume. The biggest jumps occurred in 2011–2012 (+45.9%) and 2022–2023 (+20.4%) — both tied to major regulatory and operational shifts.

💳 **By Report Type:**

EFT (Electronic Funds Transfers): ~805M reports (≈59%) — dominant and steadily rising.

LCTR (Large Cash Transactions): ~396M reports (≈29%) — declining after 2020 due to reduced cash activity.

STR (Suspicious Transactions): ~151M reports (≈11%) — doubled since 2017, highlighting enhanced AML vigilance.

CDR (Casino Disbursements): ~2M reports — rebounded after pandemic slowdowns.

🏦 **By Sector:**

Banks dominate with over 800M EFT and 396M LCTR filings — driving nearly 90% of total reports.

Money Services Businesses (MSBs) show strong STR concentration (~3M), marking a key AML focus area.

Credit Unions remain consistent contributors (~38M total).

Casinos, Real Estate, and Precious Metals sectors report smaller volumes but carry higher inherent risk.

🌎 **Geographically:**

Ontario leads (~40% of all reports), followed by Quebec (~17%) and British Columbia (~15%).

Activity is heavily concentrated in Toronto, Montreal, and Vancouver — Canada’s major financial hubs.

📅 **Fiscal Trend:**

Total annual reports rose from ~68M in FY2011–12 to ~180M in FY2022–23.

EFTs and STRs are driving this upward curve, underscoring a clear shift toward digital vigilance and risk-based reporting.

In short:

Canada’s AML reporting landscape is maturing rapidly — from cash-heavy monitoring to sophisticated, technology-driven oversight.

The continued rise in STRs and EFTs highlights growing vigilance and stronger compliance systems across the industry.

This analysis provides insights into the reporting patterns captured by FINTRAC data, highlighting trends over time, key geographical areas, and the distribution of different report types.